lib/api_browser/app/js/factories/template_for.js

```diff
@@ -0,0 +1,110 @@
+ * $family             | String  | The family or group of types this type belongs to.
+ * $typeDefinition     | Object  | The full definition of the type as available.
+ * $requestedTemplate  | String  | This is the kind of template that we want. Possible values are: standalone, embedded and label. Embedded templates represent types that are attributes of another type. The template thus should assume that it is being rendered in a table row. Label templates display a human readable name of a type. Standalone templates are used when a type is displayed by itself (i.e. a request payload).
+ *
+ * The type resolver should return either a $compiled template, or a url of a
+ * template or a promise for one of those. In that case the template resolution
+ * procedure will halt and the returned template will be displayed. If `undefined`
+ * is passed, the next resolver will be invoked. A promise for undefined is currently
+ * not supported.
+ */
+app.provider('templateFor', function() {
+
+  var resolvers = [];
+  /**
+   * Register a custom template resolver.
+   * @param {function} resolver
+   *    A function that returns undefined or a promise for a compiled template or a template url.
+   */
+  this.register = function(resolver) {
+    resolvers.unshift(resolver);
+  };
+
+  this.register(/*ngInject*/function payloadResolver($family, $requestedTemplate) {
+    if ($requestedTemplate === 'standalone') {
+      switch ($family) {
+        case 'Struct':
+          return 'views/types/standalone/struct.html';
+        default:
+          return 'views/types/standalone/default.html';
+      }
+    }
+  });
+
+  this.register(/*ngInject*/function typeResolver($family, $type, $requestedTemplate) {
+    if ($requestedTemplate === 'embedded') {
+      if ($type === 'Links') {
+        return 'views/types/embedded/links.html';
+      }
+      switch ($family) {
+        case 'Struct':
+          return 'views/types/embedded/struct.html';
+        default:
+          return 'views/types/embedded/default.html';
+      }
+    }
+  });
+
+  this.register(/*ngInject*/function labelResolver($typeDefinition, $requestedTemplate, primitives) {
+    if ($requestedTemplate === 'label') {
+      if (_.contains(primitives, $typeDefinition.name)) {
+        return 'views/types/label/primitive.html';
+      } else if ( $typeDefinition.member_attribute !== undefined ) {
+        if ( _.contains(primitives, $typeDefinition.member_attribute.type.name)){
+          return 'views/types/label/primitive_collection.html';
+        } else{
+          return 'views/types/label/type_collection.html';
+				}
+      } else if ($typeDefinition.link_to) {
+        return 'views/types/label/link.html';
+      }
+      return 'views/types/label/type.html';
+    }
+  });
+
+  this.$get = function($q, $http, $injector, $compile, $templateCache) {
+    return function(type, templateType) {
+      for (var i = 0; i < resolvers.length; i++) {
```

Comment:
Because I want to terminate the iteration early if a solution is found. I guess I could use something like `_.find`, but it felt unintuitive.